overlord/devicestate/devicemgr.go

```diff
@@ -1048,6 +1051,72 @@ func (m *DeviceManager) ensureSeedInConfig() error {
 
 }
 
+func (m *DeviceManager) appendTriedRecoverySystem(label string) error {
+	// state is locked by the caller
+
+	var triedSystems []string
+	if err := m.state.Get("tried-systems", &triedSystems); err != nil && err != state.ErrNoState {
+		return err
+	}
+	if strutil.ListContains(triedSystems, label) {
+		// system already recorded as tried?
+		return nil
+	}
+	triedSystems = append(triedSystems, label)
+	m.state.Set("tried-systems", triedSystems)
+	return nil
+}
+
+func (m *DeviceManager) ensureTriedRecoverySystem() error {
+	if release.OnClassic {
+		return nil
+	}
+	if m.systemMode != "run" {
+		return nil
+	}
+	if m.ensureTriedRecoverySystemRan {
+		return nil
+	}
+
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	deviceCtx, err := DeviceCtx(m.state, nil, nil)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	outcome, label, err := boot.InspectTryRecoverySystemOutcome(deviceCtx)
```

Comment:
now that we are looking at the full circle: we went -> recover -> back to run mode here,  this helper should check that label belongs as well to modeenv CurrentRecoverySystems and not report success unless that's the case I think but is not doing that atm afaict.

For the rest this code seems correct to

Reply:
yeah agreed, this doesn't seem to be doing that, I can try to quickly push a separate patch enabling that behavior

Reply:
I pushed a patch for this here, hopefully it's in the right direction to unblock this